keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1149 @@
+prone to this sort of error. And although Kubernetes consistently decodes JSON
+objects containing duplicate keys, the presence of duplicate keys indicates a
+mistake. Protobuf is typically machine-generated, and decoders are expected to
+be “last one wins” in the case of duplicated fields. So while it would be
+unexpected for a Protobuf-encoded object to contain duplicate fields, the
+interpretation of such an object is unambiguous.
+
+A map containing duplicate keys is well-formed but invalid according to the CBOR
+specification. Decoding a map containing duplicate keys will produce a decode
+error.
+
+<<[UNRESOLVED]>>
+How should the decoder handle unrecognized fields, that is, encountering a map
+key that does not correspond to the name of a struct field's `json` tag name?
+This may be expected in cases where the client is newer than the server, or
+where an object containing an unrecognized field was transcoded from YAML or
+JSON to CBOR. A strict decode error (as in JSON) may be appropriate in this case.
+
+This is not a concern for the custom resource path, which uses a schema-aware
+decoder wrapper to detect unknown fields.
+<<[/UNRESOLVED]>>
+
+Note that clients (e.g. kubectl) may choose to decode an object from a JSON or
+YAML text representation containing duplicate keys, then encode to CBOR to
+populate the body of an API request. Since the text-encoded content (potentially
+containing duplicates) is not literally transcoded to CBOR, this use case is
+supported. Depending on the strictness mode, duplicate keys would either be
+removed or produce an error at decode time.
+
+### Encoding Determinism
+
+It is possible for a single object to be encodable as multiple distinct but
+valid and semantically-equivalent CBOR byte strings. The CBOR specification does
+not require encoder implementations to produce deterministic output, although it
+does include recommendations for implementing deterministic encoding.
+
+The etcd3 storage implementation of GuaranteedUpdate relies on deterministic
+encoding to skip writes if the stored bytes would not change. The existing JSON
+and Protobuf encoders produce deterministic output.
+
+Encode benchmarks for the two evaluated Go CBOR libraries show a 2.4x speedup
+and a 1.8x speedup by disabling map key sorting. According to the spec, “the
+CBOR data model for maps does not allow ascribing semantics to the order of the
+key/value pairs in the map representation.”  And since the CBOR decoder will
+reject maps containing duplicate keys, a CBOR map represents exactly the same
+set of key-value pairs regardless of the order they are encoded.
+
+In order to take advantage of the available speedup, the CBOR encoder will
+support separate deterministic and nondeterministic modes. The deterministic
+mode will be used for storage serialization only. The nondeterministic mode
+should actively randomize the order of map item encoding (as with map iteration
```

Comment:
I agree we don't want clients to assume an order for maps. All our deserializers (client and server) absolutely must be able to read a map in any order.

Is there a cost to us to randomize the order for non-storage serialization?

Reply:
It depends. I'm not aware of a way to randomly iterate through the entries of a Go map that is both safe and faster than with a range expression.

For encoding from a Go struct, both of the implementations I've looked at precompute an internal list of field descriptions per type and iterate over that during encoding. I don't know how expensive it would be to choose a random field permutation every time a struct is encoded. If the cost is significant, maybe something less than true randomness would achieve the same goal, like starting iteration from a random offset or sometimes in reverse?

Reply:
If this turns out to be cheap, then I'd say do what is proposed. If there is a non-negligible cost, maybe we could enable randomness during some tests, but let production do whatever is fastest?

Reply:
I would be uncomfortable with making it configurable. Someone developing an operator and testing it end-to-end (for example) is unlikely to realize that they should enable randomness and any assumptions they're making about item order are more likely to go unnoticed. I updated the text to state that this shouldn't come at a significant performance cost, and we can discuss again if it turns out to be an obstacle. Does that sound OK?